Added Create Employee Domain and treated some exceptions

topleansebastian opened the pull request:
- Create Employee Domain added
- Employee class added
- Added some cases for a restaurant to not be successfully created and treated them in Program.cs

topleansebastian committed:
Added Create Employee Domain and treated some exceptions

- Create Employee Domain added
- Employee class added
- Added some cases for a restaurant to not be successfully created and treated them in Program.cs

cosmintiru commented on Toplean_Alexandru/week-02/Demo/Program.cs:
```diff
@@ -0,0 +1,125 @@
+            else
+            {
+                // Treating all the possible errors
+                var resultClass = result as RestaurantNotCreated;
+                Console.Write("Error! Restaurant was not created because the ");
+                switch (resultClass.Reason)
+                {
+                    case RestaurantErrorCode.NameTooLong:
+                        {
+                            Console.WriteLine($"restaurant name was too long. Please use a name shorter than 256 characters! Error code {resultClass.Reason}");
+                            break;
+                        }
+                    case RestaurantErrorCode.IllegalCharacters:
+                        {
+                            Console.WriteLine($"restaurant name had illegal characters. Error code {resultClass.Reason}");
+                            break;
+                        }
+                    case RestaurantErrorCode.RestaurantExists:
+                        {
+                            Console.WriteLine($"restaurant already exists in the system. Error code {resultClass.Reason}");
+                            break;
+                        }
+                    case RestaurantErrorCode.EmptyField:
+                        {
+                            Console.WriteLine($"restaurant name was empty. Error code {resultClass.Reason}");
+                            break;
+                        }
+                }
+            }
+
+            var restaurantCreated = result as RestaurantCreated; // Assuming we have the createdRestaurant or using an existing restaurant
+
+            // Trying to create a menu
```
I think that this is a very good start to understand the pattern matching properties of that result. What happened here is that you actually tried to avoid the fact that we are hiding all the useful data behind that abstraction (ICreateRestaurantResult). 

Would you please try to refactor this code using the Match extension methods? You already have them defined and invoked in the following line:

```csharp
  var finalResult = result.Match<bool>(OnRestaurantCreated, OnRestaurantNotCreated);
```
...you just have to fill in the implementations for OnRestaurantCreated and OnRestaurantNotCreated.

cosmintiru commented on Toplean_Alexandru/week-02/Domain/Domain/CreateRestauratOp/CreateRestaurantOp.cs:
```diff
@@ -0,0 +1,44 @@
+        {
+            //validate
+
+            return Exists(Op.Name) ?
```
Would it be fair to assume that most of these validations belong to the actual command? 

From my point of view the command itself should contain the validation logic which should be called from the operation itself. 

Please be aware that that does not include the !"Exists" preconditions. That precondition checks that a given restaurant exists in your environment, which is not in the ownership of the command. 

Can you try a refactoring to clearly separate the responsibilities?

topleansebastian committed:
Update Program.cs

topleansebastian committed:
Updated the Restaurant code to meet the requirements

- Treated the "UnknownError" case
- The members in the Cmd are checked in the same file by the IsValid() function
- Changed CreateRestaurantOP file
- Added some new ErrorCodes in Restaurant.cs

topleansebastian committed:
Partially done Menu Domain

topleansebastian committed:
Mostly finished CreateMenu and CreateRestaurant Domain

- Mostly finished CreateMenu and CreateRestaurant Domain
- Started working on CreateEmployee Domain
- Created the files for SelectRestaurant Domain
- Updated the code in Program.cs

topleansebastian committed:
Added CreateEmployee and CreateOrder Domain

- Added CreateEmployeeOp
- Added CreateOderOp
- Updated Restaurant model for storing the Employee and Order aswell

diercan commented on Toplean_Alexandru/week-02/Domain/Domain/CreateEmployeeOp/CreateEmployeeOp.cs:
```diff
@@ -0,0 +1,35 @@
+                return Task.FromResult<ICreateEmployeeResult>(new EmployeeNotCreated(EmployeeErrorCode.NameExists)); // Employee already exists
+            else
+            {
+                if (Op.IsValid().Item1)
```
Please use deconstruction instead of accessing elements in tuple via the Item properties: https://docs.microsoft.com/en-us/dotnet/csharp/deconstruct. Code is much more clear when you use deconstruction.

diercan commented on Toplean_Alexandru/week-02/Domain/Domain/CreateMenuOp/CreateMenuCmd.cs:
```diff
@@ -0,0 +1,51 @@
+                    Restaurant == null ? (false, MenuErrorCode.NullRestaurant) :
+                    (true, MenuErrorCode.None);
+            }
+            catch
```
When you catch an exception please make sure you record the caught exception by doing ToString on it, this way you will not loose useful information, in this case I would not return just UnknownError because this will not be useful when you have to debug the problem, instead I would prefer that the IsValid method will return the boolean, that tells if the command is valid or not and a string which contains the error message if the input is not valid, and if there was an exception I expect  to have there the exception ToString

diercan commented on Toplean_Alexandru/week-02/Domain/Domain/CreateMenuOp/CreateMenuCmd.cs:
```diff
@@ -0,0 +1,51 @@
+        {
+            try
+            {
+                return HasIllegalCharacters(Name) ? (false, MenuErrorCode.IllegalCharacter) :
```
Please only use conditional ternary operator (?:) when you want to evaluate a single condition, don't use imbrication constructions of ternary operator, it is error prone and can generate problems, in this case the old if-else statement is clear and easier to read

diercan commented on Toplean_Alexandru/week-02/Domain/Domain/CreateMenuOp/CreateMenuCmd.cs:
```diff
@@ -0,0 +1,51 @@
+
+        public bool HasIllegalCharacters(String name)
+        {
+            foreach (char c in name)
```
Try to rewrite this check using Any LINQ operator: https://docs.microsoft.com/en-us/dotnet/api/system.linq.enumerable.any?view=netframework-4.8

diercan commented on Toplean_Alexandru/week-02/Domain/Domain/CreateMenuOp/CreateMenuOp.cs:
```diff
@@ -0,0 +1,37 @@
+                return Task.FromResult<ICreateMenuResult>(new MenuNotCreated(MenuErrorCode.ExistentMenu)); // Menu already exists
+            else
+            {
+                if (Op.IsValid().Item1)
```
Use deconstruction.

topleansebastian committed:
Added SelectRestaurant Domain and fixed bad code

- Removed foreach loop with Any() method
- Replaced short ifs with the standard if else where needed
- Deconstructed all tuples
- Catch block will return the exp.ToString() instead of "Unknown Error"
- Added SelectRestaurant Domain
-  Updated Program.cs and commented the SelectRestaurant expression.
- Select restaurant will return an Restaurant Object from the hardcoded list rather than creating a new instance

topleansebastian committed:
Added DeleteRestaurant and CreateMenuItem Domains

- Added DeleteRestaurant that returns bool if it was successfully deleted
- Added CreateMenuItem and binded the MenuItem object to Menu.
- Bugfix: Created a single instance of Menu and used it for all bindings rather than creating an individual instance of Menu object for each bind operation
